File: ios/Classes/PublicPresentationPlugin.swift
import Flutter
import UIKit
import SuperwallKit

public class PublicPresentationPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "SWK_PublicPresentationPlugin", binaryMessenger: registrar.messenger())
    let instance = PublicPresentationPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }
  
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
      case "dismiss":
        Task {
          await Superwall.shared.dismiss()
          result(nil)
        }
        
      case "registerEventWithFeature":
        guard let args = call.arguments as? [String: Any],
              let event = args["event"] as? String else {
          result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for registerEventWithFeature", details: nil))
          return
        }
        
        let params = args["params"] as? [String: Any]

        Superwall.shared.register(event: event, params: params) {
          // TODO: Handle state updates
        }
        result(nil)
        
      case "registerEvent":
        guard let args = call.arguments as? [String: Any],
              let event = args["event"] as? String else {
          result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for registerEvent", details: nil))
          return
        }

        let params = args["params"] as? [String: Any]

        Superwall.shared.register(event: event, params: params, handler: nil)
        result(nil)
        
      default:
        result(FlutterMethodNotImplemented)
    }
  }
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/PublicPresentationPlugin.kt
package com.superwall.superwallkit_flutter

import com.superwall.sdk.Superwall
import com.superwall.sdk.paywall.presentation.dismiss
import com.superwall.sdk.paywall.presentation.register
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class PublicPresentationPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private val scope = CoroutineScope(Dispatchers.Main)

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "SWK_PublicPresentationPlugin")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "dismiss" -> {
                scope.launch {
                    Superwall.instance.dismiss()
                    result.success(null)
                }
            }
            "registerEventWithFeature" -> {
                val event = call.argument<String>("event")
                if (event != null) {
                    val params = call.argument<Map<String, Any>>("params")
//                    val handler = call.argument<Int>("handler")

                    // TODO: Implement the logic to register an event in Superwall with a feature handler
                    Superwall.instance.register(event, params)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENTS", "Invalid arguments for registerEventWithFeature", null)
                }
            }
            "registerEvent" -> {
                val event = call.argument<String>("event")
                if (event != null) {
                    val params = call.argument<Map<String, Any>>("params")
                    Superwall.instance.register(event, params)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENTS", "Invalid arguments for registerEvent", null)
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
